projects/extension/src/background/ConnectionManager.ts

```diff
@@ -1,297 +1,425 @@
+                    (untrustedChainId): SmoldotChain[] => {
+                      const chain = sandbox.chains.get(untrustedChainId)
+                      return chain && chain.isReady ? [chain.smoldotChain] : []
+                    },
+                  )
+                : [],
+          })
+
+        // Insert the promise in `sandbox.chains` so that the state machine is aware of the
+        // fact that there is a chain with this ID currently initializing.
+        const name = nameFromSpec(chainSpec)
+        sandbox.chains.set(message.chainId, {
+          isReady: false,
+          smoldotChain: chainInitialization,
+          name,
         })
-    })
 
-    const chain = await chainPromise
+        // Spawn in the background an async function that is called once the initialization
+        // finished, either successfully or not.
+        ;(async () => {
+          // `result` contains either the chain (on success) or the error message (on failure).
+          let result: SmoldotChain | string
+          try {
+            result = await chainInitialization
+          } catch (err) {
+            result =
+              err instanceof Error
+                ? err.message
+                : "Unknown error when adding chain"
+          }
+
+          // Because the chain initialization might have taken a long time, we first need to
+          // check whether the chain that we're initializing is still in `this`, as it might
+          // have been removed by various other functions if it no longer interests us.
+          const sandbox = this.#sandboxes.get(sandboxId)
+          if (
+            !sandbox ||
+            !(sandbox.chains.get(chainId)?.smoldotChain === chainInitialization)
+          ) {
+            typeof result !== "string" && result.remove()
+            return
+          }
+
+          // Update `sandbox.chains`, either updating the entry on success or removing it on
+          // failure, and send back a message.
+          if (typeof result !== "string") {
+            const smoldotChain: SmoldotChain = result
+            healthChecker.setSendJsonRpc((rq) => smoldotChain.sendJsonRpc(rq))
+            const readyChain: ReadyChain = {
+              isReady: true,
+              name,
+              smoldotChain,
+              healthChecker,
+            }
+            healthChecker.start((health) => {
+              readyChain.latestHealthStatus = health
+            })
+            sandbox.chains.set(chainId, readyChain)
+            sendSandbox(sandbox, {
+              origin: "substrate-connect-extension",
+              type: "chain-ready",
+              chainId: message.chainId,
+            })
+          } else {
+            sandbox.chains.delete(chainId)
+            sendSandbox(sandbox, {
+              origin: "substrate-connect-extension",
+              type: "error",
+              chainId: message.chainId,
+              errorMessage: result,
+            })
+          }
+        })()
```

Comment:
A bit of a nit-pick, but I would much prefer having the error handling logic inside the `catch` clause, rather than the `if/else`s for a result that can have either the error message or the chain.

```suggestion
        ;(async () => {                                                         
          let sandbox: Sandbox | undefined                                      
          try {                                                                 
            const smoldotChain = await chainInitialization.finally(() => {      
              sandbox = this.#sandboxes.get(sandboxId)                          
            })                                                                  
                                                                                
            if (                                                                
              sandbox?.chains.get(chainId)?.smoldotChain !== chainInitialization
            ) {                                                                 
              smoldotChain.remove()                                             
              return                                                            
            }                                                                   
            healthChecker.setSendJsonRpc((rq) => smoldotChain.sendJsonRpc(rq))  
            const readyChain: ReadyChain = {                                    
              isReady: true,                                                    
              name,                                                             
              smoldotChain,                                                     
              healthChecker,                                                    
            }                                                                   
            healthChecker.start((health) => {                                   
              readyChain.latestHealthStatus = health                            
            })                                                                  
            sandbox.chains.set(chainId, readyChain)                             
            sendSandbox(sandbox, {                                              
              origin: "substrate-connect-extension",                            
              type: "chain-ready",                                              
              chainId: message.chainId,                                         
            })                                                                  
          } catch (err) {                                                       
            if (!sandbox) return                                                
            sandbox.chains.delete(chainId)                                      
            sendSandbox(sandbox, {                                              
              origin: "substrate-connect-extension",                            
              type: "error",                                                    
              chainId: message.chainId,                                         
              errorMessage:                                                     
                err instanceof Error                                            
                  ? err.message                                                 
                  : "Unknown error when adding chain",                          
            })                                                                  
          }                                                                     
        })()                                                                    
```

Reply:
Just want to point out that it's not the same code. Re-grabbing the sandbox from `this` is intentionally done after the `await`. The comment explains that.

Reply:
> Re-grabbing the sandbox from this is intentionally done after the await

So it is in the code that I'm suggesting.